demoapp/extract_mgl_sections.py

```diff
@@ -0,0 +1,319 @@
+            #iterate over list that contains only sections or chapters
+            for i, j in enumerate(x):
+                #ignore SECTION with caps as it indicates sections from the bills and not the MGL
+                if j == "SECTION": 
+                    continue
+                if j == "":
+                    continue
+                
+                else:
+                    if j.casefold() == "chapter":
+                        current_chapter = x[i+1].strip() #keep track of current chapter
+                    
+                    if j.casefold() == "section":
+                        if x[i+1] == "":
+                            continue
+                        else:
+                            section = x[i+1].strip()
+                        
+                        list_with_both = [current_chapter, section]
+                        if list_with_both not in lists_with_both:
+                            lists_with_both.append(list_with_both)
+                            
+    return lists_with_both
+
+def apply_extract_sections(bills_df, bill_text_col_name, chap_sec_lists_col_name):
+    
+    """
+    Applies the 'extract_sections' function to a DataFrame containing the bills text
+
+    This function adds a new column 'Reference List' to the input DataFrame 'bills_df'. The 'DocumentText'
+    column of 'bills_df' is processed using the 'extract_sections' function, and the results are stored in
+    the newly created 'Reference List' column.
+
+    Parameters:
+    - bills_df (pd.DataFrame): The DataFrame containing bill information with a 'DocumentText' column.
+
+    Returns:
+    - bills_df (pd.DataFrame): The updated DataFrame with a new 'Reference List' column.
+    """
+
+    
+    bills_df [chap_sec_lists_col_name] = bills_df[bill_text_col_name].apply(extract_sections)
+    
+    return bills_df
+
+def make_api_call(chapter_section_list):
+    """
+    Makes an API call to retrieve text data based on the provided chapter and section.
+
+    Parameters:
+    - chapter_section_list (list): A list containing two elements - chapter and section, e.g., ['2', '15D'].
+
+    Returns:
+    - result (str): The text data retrieved from the API.
+
+    Note:
+    - This function uses the malegislature.gov API to fetch text data for a specific chapter and section.
+    
+    """
+    
+    result = """"""
+  
+    try:
+        # unpack section and chapter for example: ['2', '15D']
+        chapter, section = chapter_section_list
+        link = f'https://malegislature.gov/api/Chapters/{chapter}/Sections/{section}'
+        r = requests.get(link, verify=False)
+        r = r.json()
+
+        # fields to extract
+        result =  r.get("Text", np.nan)
+        # logging.info('API call successful')
+        return result
+    except RequestException as e:
+        # logging.error('API Request failed:section not found %s',chapter_section_list)
+        pass
+    
+
+def get_mgl_from_api(chapter_section_lists):
```

Comment:
Recommend renaming to `query_section_text_all_bills` or similar.

```suggestion
def query_section_text_all_bills(chapter_section_lists: list[tuple[str, str]]) -> tuple[list[str], list[tuple[str, str]]]:
```